### PR TITLE
[FLINK-3609] [tableAPI] Reorganize selection of optimization rules

### PR DESCRIPTION
**Pull request by fhueske**
- Remove the join reordering rules to keep the user specified join order (rules can be added if enough statistics are available).
- Copy and fix a broken Calcite rule (rule will be fixed with Calcite 1.7).
- Add two test cases.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala [hunk 1]
```diff
@@ -29,50 +29,45 @@ object FlinkRuleSets {
     */
   val DATASET_OPT_RULES: RuleSet = RuleSets.ofList(
 
-    // filter rules
+    // push a filter into a join
     FilterJoinRule.FILTER_ON_JOIN,
+    // push filter into the children of a join
     FilterJoinRule.JOIN,
-    FilterMergeRule.INSTANCE,
```
Comment (vasia) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala]: Why don't we need this one?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala [hunk 2]
```diff
@@ -29,50 +29,45 @@ object FlinkRuleSets {
...
-    FilterAggregateTransposeRule.INSTANCE,
+    // push filter through an aggregation
+    FlinkFilterAggregateTransposeRule.INSTANCE,
 
-    // push and merge projection rules
+    // aggregation and projection rules
     AggregateProjectMergeRule.INSTANCE,
-    ProjectMergeRule.INSTANCE,
```
Comment (vasia) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala]: And this one? Are we covered by the CalcMergeRule?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala [hunk 3]
```diff
@@ -29,50 +29,45 @@ object FlinkRuleSets {
...
 
-    // merge and push unions rules
-    // TODO: Add a rule to enforce binary unions
+    // join rules
+    JoinPushExpressionsRule.INSTANCE,
+
+    // remove union with only a single child
     UnionEliminatorRule.INSTANCE,
-    FlinkJoinUnionTransposeRule.LEFT_UNION,
-    FlinkJoinUnionTransposeRule.RIGHT_UNION,
```
Comment (vasia) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala]: Why did you remove these 2?

**Comment by vasia** [comment]
I have some doubts about the filter/project merge rules and the `FlinkJoinUnionTransposeRule`. Otherwise looks good!

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala (hunk 1 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala]: Should be covered by the rules to merge Calcs.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala (hunk 2 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala]: Should be covered by the rules to merge Calcs.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala (hunk 3 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala]: These rules would push a join into a union. This means that a join is executed on each input of the union and the join results are unioned afterwards. Additional joins are likely to cause higher resource consumption. For example if the unioned inputs are fed into the probe-side of a join, the build side would need to be replicated. 

Unless we have very good statistics and a more fine-grained cost model, I would not try to optimize and trust the user (at least for the Table API).

Comment (vasia) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala]: Alright. Then `FlinkJoinUnionTransposeRule` can be removed since it's not used. I can do that before merging.

Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala]: True, will update the PR.

**Comment by fhueske** [comment]
Removed `FlinkJoinUnionTransposeRule` and updated the PR.

**Comment by vasia** [comment]
Thanks, will merge.

**Comment by vasia** [comment]
and merged!
